gdal/frmts/jpegxr/jpegxrdataset.cpp

```diff
@@ -0,0 +1,1202 @@
+    }
+
+public:
+    float fQuality = 1.0;
+
+private:
+    Quality eQuality = QUALITY_UNSET;
+    Overlap eOverlap = OVERLAP_UNSET;
+    Subsampling eSubsampling = SUBSAMPLING_UNSET;
+};
+
+/************************************************************************/
+/* ==================================================================== */
+/*                         JPEGXREncoder                                */
+/* ==================================================================== */
+/************************************************************************/
+
+struct JPEGXREncoder final
+{
+    friend struct JPEGXREncoderConfig;
+
+    JPEGXREncoder()
+        : pFactory(nullptr)
+        , pCodecFactory(nullptr)
+        , pImageEncode(nullptr)
+        , pEncodeStream(nullptr)
+    {
+    }
+
+    ~JPEGXREncoder()
+    {
+        if (pImageEncode)
+            pImageEncode->Release(&pImageEncode); // also releases attached stream
+        pImageEncode = nullptr;
+
+        if (pCodecFactory)
+            pCodecFactory->Release(&pCodecFactory);
+        pCodecFactory = nullptr;
+
+        if (pFactory)
+            pFactory->Release(&pFactory);
+        pFactory = nullptr;
+    }
+
+    void Initialize(char const* pszFilename)
+    {
+        CPLAssert(!pFactory);
+        CPLAssert(!pCodecFactory);
+        CPLAssert(!pImageEncode);
+        ERR nErr = WMP_errFail;
+
+        /* -------------------------------------------------------------------- */
+        /*      Create encoding stream                                          */
+        /* -------------------------------------------------------------------- */
+        nErr = PKCreateFactory(&pFactory, PK_SDK_VERSION);
+        if (nErr != WMP_errSuccess || !pFactory)
+            throw JPEGXRException(nErr);
+
+        nErr = pFactory->CreateStreamFromFilename(&pEncodeStream, pszFilename, "wb");
+        if (nErr != WMP_errSuccess || !pEncodeStream)
+            throw JPEGXRException(nErr);
+
+        /* -------------------------------------------------------------------- */
+        /*      Create encoder                                                  */
+        /* -------------------------------------------------------------------- */
+        nErr = PKCreateCodecFactory(&pCodecFactory, WMP_SDK_VERSION);
+        if (nErr != WMP_errSuccess || !pCodecFactory)
+            throw JPEGXRException(nErr);
+
+        nErr = pCodecFactory->CreateCodec(&IID_PKImageWmpEncode,
+                                          reinterpret_cast<void**>(&pImageEncode));
+        if (nErr != WMP_errSuccess || !pImageEncode)
+            throw JPEGXRException(nErr);
+
+        /* -------------------------------------------------------------------- */
+        /*      Set default encoding parameters                                 */
+        /* -------------------------------------------------------------------- */
+        CPLAssert(pImageEncode);
```

Comment:
@rouault yes, they only serve as a comment. And, in case someone rearranges the above code without reviewing the below code...in future.